### PR TITLE
Remove clustermanagementaddon CR search-collector if search is removed

### PR DESCRIPTION
**Pull request by xrajesh**
**Related Issue:**  stolostron/backlog#<ISSUE_NUMBER>

### CODE REVIEW

**Comment by zhiweiyin318** [comment]
I have some confusion why delete `clusterManagementAddon` when search CR is deleted?
`clusterManagementAddon` on the hub cluster is to say search-collector addon operator (search-v2-operator) has been installed on the hub, you can enable the search-collector addon agent on your managed cluster by search-collector managedClusterAddon CR.
I think `clusterManagementAddon` should be deleted only when the search-collector addon operator (search-v2-operator) is removed.

**Comment by xrajesh** [comment]
@zhiweiyin318 Thank you for your feedback. We are creating `clusterManagementAddon` instance named `search-collector` in the Reconcile function , which executes only after a `search` instance named `search-v2-operator` is created. So once we remove the search instance, we can also remove  `clusterManagementAddon` instance named `search-collector` right , because - next time a search CR instance is created we create the  `clusterManagementAddon`  again. Currently we are creating the search-v2-operator from a bundle, and the search-v2-operator does not do anything until we have a search CR @jlpadilla FYI.

File: controllers/search_controller.go [hunk 1]
```diff
@@ -180,6 +191,48 @@ func (r *SearchReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+			// that we can retry during the next reconciliation.
+			if err := r.finalizeSearch(instance); err != nil {
+				return err
+			}
+			// Remove searchFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			controllerutil.RemoveFinalizer(instance, searchFinalizer)
+			r.Update(ctx, instance)
+			log.Info("Finalizer removed from search CR")
+		}
```
Comment (zhiweiyin318) [inline controllers/search_controller.go:226]: should return here ?

Comment (xrajesh) [reply controllers/search_controller.go:230]: Done .

File: controllers/search_controller.go [hunk 2]
```diff
@@ -62,17 +64,30 @@ func (r *SearchReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
...
 	if IsPaused(instance.GetAnnotations()) {
 		log.Info("Reconciliation is paused because the annotation 'search-pause: true' was found.")
 		return ctrl.Result{}, nil
 	}
+
+	// Setup finalizers
+	deleted, err := r.setFinalizer(ctx, instance)
+	if err != nil || deleted {
+		if deleted {
+			log.V(2).Info("Search Instance deleted, reque request", err)
```
Comment (SherinV) [inline controllers/search_controller.go:82]: requeue*

Comment (SherinV) [reply controllers/search_controller.go:82]: Same on line 84

File: controllers/search_controller.go [hunk 3]
```diff
@@ -180,6 +195,50 @@ func (r *SearchReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+			// Run finalization logic for searchFinalizer. If the
+			// finalization logic fails, don't remove the finalizer so
+			// that we can retry during the next reconciliation.
+			if err := r.finalizeSearch(instance); err != nil {
+				return false, err
+			}
+			// Remove searchFinalizer. Once all finalizers have been
+			// removed, the object will be deleted.
+			controllerutil.RemoveFinalizer(instance, searchFinalizer)
+			r.Update(ctx, instance)
```
Comment (SherinV) [inline controllers/search_controller.go:228]: log and return error if update fails?

File: controllers/cleanup.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+			Name:      "search-collector",
+			Namespace: instance.GetNamespace(),
+		},
+	}
+	err := r.Delete(context.TODO(), cma)
+	if err != nil && !errors.IsNotFound(err) {
+		log.Error(err, "Failed to delete ClusterManagementAddon", "name", cma)
+		return err
+	}
+	time.Sleep(1 * time.Second)
```
Comment (jlpadilla) [inline controllers/cleanup.go:31]: Why do we need to sleep?

Comment (jlpadilla) [reply controllers/cleanup.go:31]: If it's for the retry logic, maybe only sleep if there's and error?

Comment (xrajesh) [reply controllers/cleanup.go:31]: Had that only as a safety measure - Did not want to start a reconcile on a instance which is SET to delete .I am removing it for now , if we see any issues , we can check again.

**Review by SherinV — APPROVED** [review]

**Review by jlpadilla — APPROVED** [review]
/lgtm
/hold

**Pull request merged**